### PR TITLE
codec/json: add u64 support.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/util/codec/number.rs [hunk 1]
```diff
@@ -126,6 +126,10 @@ pub trait NumberEncoder: Write {
     fn encode_i64_le(&mut self, v: i64) -> Result<()> {
         self.write_i64::<LittleEndian>(v).map_err(From::from)
     }
+
+    fn encode_u64_le(&mut self, v: u64) -> Result<()> {
```
Comment (AndreMouche) [inline src/util/codec/number.rs]: any test?

File: src/util/codec/number.rs [hunk 2]
```diff
@@ -204,6 +208,10 @@ pub trait NumberDecoder: Read {
     fn decode_i64_le(&mut self) -> Result<i64> {
         self.read_i64::<LittleEndian>().map_err(From::from)
     }
+
+    fn decode_u64_le(&mut self) -> Result<u64> {
```
Comment (AndreMouche) [inline src/util/codec/number.rs]: Please add unit test for this function

File: src/coprocessor/codec/mysql/json/serde.rs [hunk 3]
```diff
@@ -64,6 +64,7 @@ impl Serialize for Json {
             }
             Json::Double(d) => serializer.serialize_f64(d),
             Json::I64(d) => serializer.serialize_i64(d),
+            Json::U64(d) => serializer.serialize_u64(d),
```
Comment (AndreMouche) [inline src/coprocessor/codec/mysql/json/serde.rs]: Please adjust the unit test to cover both `Json::u64` and `Json::i64`

**Comment by AndreMouche** [comment]
I think you should adjust the source datum.rs like https://github.com/pingcap/tikv/blob/master/src/coprocessor/codec/datum.rs#L262 
and tests in file https://github.com/pingcap/tikv/blob/master/src/coprocessor/xeval/evaluator.rs too

File: src/coprocessor/codec/mysql/json/comparison.rs [hunk 4]
```diff
@@ -37,7 +37,7 @@ impl Json {
             Json::Array(_) => PRECEDENCE_ARRAY,
             Json::Boolean(_) => PRECEDENCE_BOOLEAN,
             Json::None => PRECEDENCE_NULL,
-            Json::I64(_) | Json::Double(_) => PRECEDENCE_NUMBER,
+            Json::I64(_) | Json::U64(_) | Json::Double(_) => PRECEDENCE_NUMBER,
```
Comment (hhkbp2) [inline src/coprocessor/codec/mysql/json/comparison.rs]: Add test to compare U64?

**Comment by hhkbp2** [comment]
LGTM

File: src/util/codec/number.rs (hunk 2 above)
Comment (hicqu) [inline src/util/codec/number.rs]: fixed.

File: src/coprocessor/codec/mysql/json/comparison.rs (hunk 4 above)
Comment (hicqu) [inline src/coprocessor/codec/mysql/json/comparison.rs]: fixed.

**Comment by hicqu** [comment]
@AndreMouche @hhkbp2 , PTAL, thanks.

**Comment by hicqu** [comment]
@AndreMouche @hhkbp2 , PTAL, thanks.

**Comment by hhkbp2** [comment]
LGTM

**Comment by andelf** [comment]
What's the consideration behind this PR?

**Comment by AndreMouche** [comment]
I think you have missed function in https://github.com/pingcap/tikv/blob/master/src/coprocessor/codec/datum.rs#L429 @hicqu

File: src/coprocessor/codec/mysql/json/binary.rs [hunk 5]
```diff
@@ -407,7 +420,9 @@ mod test {
         let legal_cases = vec!{
             (r#"{"key":"value"}"#, TYPE_CODE_OBJECT),
             (r#"["d1","d2"]"#, TYPE_CODE_ARRAY),
-            (r#"3"#, TYPE_CODE_I64),
+            (r#"-3"#, TYPE_CODE_I64),
+            (r#"3"#, TYPE_CODE_U64),
+            (r#"18446744073709551615"#, TYPE_CODE_U64),
```
Comment (AndreMouche) [inline src/coprocessor/codec/mysql/json/binary.rs]: Please adjust test `test_json_binary` to cover both `Json::I64` and `Json::U64`

**Comment by AndreMouche** [comment]
No, we do not  invent a new json standard. `Json::U64` is supported in mysql while we do not implement it before. @andelf

**Comment by hicqu** [comment]
@AndreMouche @BusyJay , PTAL. thanks.

File: src/coprocessor/codec/mysql/json/binary.rs [hunk 6]
```diff
@@ -389,9 +402,11 @@ mod test {
 
         let json_nil = Json::None;
         let json_bool = Json::Boolean(true);
+        let json_int = Json::I64(30);
+        let json_uint = Json::I64(30);
```
Comment (BusyJay) [inline src/coprocessor/codec/mysql/json/binary.rs]: You mean `Json::U64(30)`?

**Comment by hicqu** [comment]
@BusyJay , PTAL again, thanks.

**Pull request merged**